Add support for classpath resources

jjudd opened the pull request:
Some third party libraries expect resources to be found on the classpath at the root of the tree. For example, as exactly "myresource.json". This commit adds support for those resources.

classpath_resources are only allowed on binaries, to avoid namespace conflicts. Just like the Bazel Java rules.

bazel-io commented:
Can one of the admins verify this patch?

jjudd commented:
When building Play framework projects we ran into an issue where Play expected things like `application.conf` to be found on the classpath as exactly `application.conf`. This is what we did to get that working.

ittaiz commented:
Thanks!
I'll take a look soon (today/tomorrow hopefully)

On Wed, Jan 31, 2018 at 6:02 AM James Judd <notifications@github.com> wrote:

> When building Play framework projects we ran into an issue where Play
> expected things like application.conf to be found on the classpath as
> exactly application.conf. This is what we did to get that working.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bazelbuild/rules_scala/pull/410#issuecomment-361817081>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABUIF8xH-Y2i8hiiUXGwsA7h3Hd9Lo9nks5tP-XggaJpZM4RzY6l>
> .
>

ittaiz commented:
And sorry for the delay

ittaiz commented:
test this please (ci bot)

ittaiz commented:
actually, what do you think about also calling your new binary in https://github.com/bazelbuild/rules_scala/blob/master/test/BUILD#L309? it makes sure scala_binary works in test environment

jjudd commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -72,10 +72,17 @@ public void processRequest(List<String> args) throws Exception {
+      /**
+       * Copy classpath resources to root of jar
+       */
+      copyClasspathResources(ops.classpathResourceFiles, tmpPath);
```
I like it. Will change.

jjudd commented on src/java/io/bazel/rulesscala/scalac/ScalacProcessor.java:
```diff
@@ -316,6 +324,26 @@ private static void copyResources(
+      if(Files.exists(target)) {
+        throw new RuntimeException(
+          "Classpath resource file " + source.getFileName()
+          + "has a namespace conflict with another file: " + target.getFileName()
```
Nice catch. Will change.

jjudd commented on test_expect_failure/classpath_resources/ObjectWithDuplicateClasspathResources.scala:
```diff
@@ -0,0 +1,5 @@
+package scala.test.classpathresources
+
+object ObjectWithDuplicateClasspathResources extends App {
+  System.exit(0)
```
I suppose not. Will change.

jjudd commented:
I'm happy to move the binaries. I'll take a look at that in the next couple days. 

I haven't thought about concatenating resources. I'll think about and get back to you.

ittaiz commented:
@jjudd I wasn't clear I'm sorry. I meant to *additionally* call your new binary in that line. The code as-is is good, the additional call is just another check

ittaiz commented:
@jjudd friendly ping

jjudd commented:
Alright. I finally had a chance to get back to this. I addressed the review comments and made classpath_resource duplicates a warning rather than an error, which is what Bazel's Java rules does: https://github.com/bazelbuild/bazel/blob/master/src/java_tools/singlejar/java/com/google/devtools/build/singlejar/SingleJar.java#L202

ittaiz commented:
Thanks!
